core/types/block_ext.go

```diff
@@ -62,7 +129,16 @@ func NewBlockWithExtData(
 	header *Header, txs []*Transaction, uncles []*Header, receipts []*Receipt,
 	hasher TrieHasher, extdata []byte, recalc bool,
 ) *Block {
-	b := NewBlock(header, txs, uncles, receipts, hasher)
-	b.setExtData(extdata, recalc)
-	return b
+	if recalc {
+		headerExtra := GetHeaderExtra(header)
+		headerExtra.ExtDataHash = CalcExtDataHash(extdata)
+	}
+	block := NewBlock(header, txs, uncles, receipts, hasher)
```

Comment:
(nit) I would move this to after the definition of `extra` as it's unused until then so there's no point in forcing the reader to parse and remember it right now.

```go
// ...
extra := {
    // ...
}
block := NewBlock(/*...*/)
extras.Block.Set(block, extra)
return block
```

Reply:
counter-nit: I see your point, although defining block then extra kind of makes it a bit more natural given the call `extras.Block.Set(block, extra)` 🤔  Or maybe that's just my 🧠 being oddly twisted 😄